### PR TITLE
Add an API to get detailed gas costs for a message

### PR DESCRIPTION
**Pull request by arajasek**

### CODE REVIEW

File: extern/storage-sealing/upgrade_queue.go [hunk 1]
```diff
@@ -67,6 +67,8 @@ func (m *Sealing) tryUpgradeSector(ctx context.Context, params *miner.SectorPreC
 		params.ReplaceSectorDeadline = loc.Deadline
 		params.ReplaceSectorPartition = loc.Partition
 
+		log.Infof("replacing sector %d with %d", *replace, params.SectorNumber)
```
Comment (arajasek) [inline extern/storage-sealing/upgrade_queue.go:70]: 100% related to this PR, i swear

**Review by magik6k — COMMENTED** [review]
Code looks good, but we may not need a separate API function - any reason why we couldn't add this to `InvocResult` from `StateReplay` / `StateCall`?

**Comment by arajasek** [comment]
@magik6k Yeah, I considered that (and we maybe should add the new fields to `InvocResult` anyway). The issue with `StateReplay` is that it needs to know which tipset the message was included in. I think we should probably have a single endpoint that both searches for the message, and upon finding it, replays it to get the gas costs.

File: node/impl/full/state.go [hunk 2]
```diff
@@ -1226,3 +1226,53 @@ func (a *StateAPI) StateCirculatingSupply(ctx context.Context, tsk types.TipSetK
...
+
+		msg = mlkp.Message
+	}
+
+	m, r, err := a.StateManager.Replay(ctx, ts, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	filLimit := big.Mul(big.NewInt(m.GasLimit), m.GasFeeCap)
```
Comment (Kubuxu) [inline node/impl/full/state.go:1267]: you can do `m.RequiredFunds()`.

Comment (arajasek) [reply node/impl/full/state.go]: Done, thanks

**Review by magik6k — APPROVED** [review]

**Review by raulk — COMMENTED** [review]
This PR introduced a regression. The regression was caught by conformance tests, but what worries me is that I did not see any discussion/mention on this PR. And the failure went ignored.

In essence it boils down to this:

image

This PR replaced the `Penalty` and `MinerTip` fields, which were always compulsory with a value of `big.Zero()` set on them if empty, by `GasCosts`. In `GasCosts` this PR only fills in the penalty, leaving _invalid_ values for the other fields of `GasCosts`:

```go
type GasOutputs struct {
	BaseFeeBurn        abi.TokenAmount
	OverEstimationBurn abi.TokenAmount

	MinerPenalty abi.TokenAmount
	MinerTip     abi.TokenAmount
	Refund       abi.TokenAmount

	GasRefund int64
	GasBurned int64
}
```

Shouldn't we be filling in all fields? Or do we want to do ugly nil checks (big.Nil()) everywhere these values are used?

**Pull request merged**